DEV: scipy.interpolate b-splines (periodic case)

diana-sukhoverkhova opened the pull request:
#### What does this implement/fix?
We developed and implemented an algorithm for computing coefficients for B-splines of degree ``k`` in periodic case.  The algorithm builds a matrix with ``k`` diagonals and upper right and lower left blocks of size ``(k-1)/2 x (k-1)/2`` (for now ``k`` is assumed to be odd). This matrix is stored in a form suitable for ``scipy.linalg.solve_banded``.  Due to the special form of matrix, it is possible to use Woodbury formula to optimize the computations. 

#### Additional information
In periodic case, ordinates and values of the first ``k-1`` derivatives are equal at the ends.

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
+        
+    Notes
+    -----
+    SLE - system of linear equations.
```
Leftover line? Can go.

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
+    SLE - system of linear equations.
+    
+    'n' should be greater than 'k', otherwise corner block
+    elements will intersect with diagonals.
```
Should be checked probably.

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
+    diagonal elements of the original matrix which makes it
+    possible to use a solution for banded matrices 
+    (``solve_banded``). Then the solution for the original SLE
+    is finally computed.
```
Can probably add the matrix structure in ASCII pseudocode here.

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
+
+    U[:bs, :bs] = ur
+    for j in range(bs): 
+        V[j, -bs + j] = 1
```
minor: can be done with fancy indexing, `V[np.arange(bs), np.arange(bs) - bs]` or some such.

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
+    References
+    ----------
+    .. [1] William H. Press, Saul A. Teukolsky, William T. Vetterling
+           and Brian P. Flannery, Numerical Recipes, 2007, 2.7.3
```
Sec 2.7.3 ?

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
+    '''
+    Implementation of Woodbury algorithm applied to banded
+    matrices with two blocks in upper right and lower left
+    corners.
```
"Solve a cyclic banded linear system with upper right and lower blocks of size `k` using the Woodbury formula" --- is this a fair description of this routine?

Then ASCII pseudocode structure of the matrix would help after the sentence of in the Notes section.

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
+
+    return c
+
+def _periodic_nodes(x,l=1,r=1):
```
nit: these are periodic knots, right?

ev-br replied:
space after comma please, here and elsewhere (until the linter CI job is happy)

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
+    n = y.shape[0]
+
+    if n <= k:
+        raise ValueError("n should be greater than k to form system.")
```
Nit: ... which system? Consider reformulating. E.g., need at least k data point to fit a spline of degree k

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
+    # set up RHS
+    A = np.zeros((k, n - 1)) # matrix of diagonals suitable for 'solve_banded'
+    for i in range(n-1):
+        A[:,i] = _bspl.evaluate_all_bspl(t, k, x[i], i + k)[:-1][::-1]
```
[::-1][::-1] looks unusual, is it really needed?

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
+        if k % 2 == 1 or i < offset:
+            A[offset + i] = np.roll(A[offset + i], -i)
+            ur[-i:, i - 1] = np.copy(A[offset + i, -i:])
+        ll[-i, :i] = np.copy(A[offset - i, :i])
```
certainly need some more docs to follow this part.

Maybe also consider writing explicit loops and move them to cython, like it's done for other boundary conditions.

ev-br commented on scipy/interpolate/tests/test_bsplines.py:
```diff
@@ -1094,59 +1151,6 @@ def make_interp_full_matr(x, y, t, k):
     return c
 
 
-### XXX: 'periodic' interp spline using full matrices
```
Why removing this? Maybe better keep this and test your implementation against.
(if this version is buggy, fix it)

ev-br reviewed (changes requested):
Initial cursory review. Did not check the algorithmic part, also `make_interp_spline` interface. Will do separately.

Full disclosure: Egor and Diana work with me at HSE, so an additional review from someone else would be nice.

ev-br commented:
When revising this, please also add tests for the number of points smaller than and equal to k, to catch edge cases similar to k=3 CubicSpline.

diana-sukhoverkhova commented:
Now this implementation works with B-splines of even degree ```k```. We changed default knot vector (in Greville-like way) and added dependency on parity to construction of matrices.

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,195 @@ def _process_deriv_spec(deriv):
+    
+    nt = len(t) - k - 1
+
+    # size of block elements
```
Consider reformulating

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,195 @@ def _process_deriv_spec(deriv):
+    _bspl._colloc(x, t, k, ab, offset=k)
+    
+    # remove zeros before the matrix
+    ab = ab[-k - (k + 1) % 2:, :]
```
Why? Consider re formulating (end of the matrix? Before the matrix?)

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,195 @@ def _process_deriv_spec(deriv):
+    # matrix while lower left is a lower triangular one.
+    for i in range(kul):
+        ur += np.diag(ab[-i - 1, i: kul], k=i)
+        ll += np.diag(ab[i, -kul - (k % 2): n - 1 + 2 * kul - i], k=-i)
```
This is due to the special choice of the knot vector, correct?

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -606,6 +607,195 @@ def _process_deriv_spec(deriv):
+
+    # remove elements that occure in the last point
+    # (first and last points are equivalent)
+    A = ab[:, kul: -k + kul]
```
Typo: occurE

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -736,37 +943,46 @@ def make_interp_spline(x, y, k=3, t=None, bc_type=None, axis=0,
+    if bc_type == 'periodic' and not np.allclose(y[0], y[-1], atol=1e-15):
+        raise ValueError("First and last points does not match while "
+                         "periodic case expected")
+
```
Identical to what CubcSpline does?

ev-br commented on scipy/interpolate/tests/test_bsplines.py:
```diff
@@ -819,6 +820,93 @@ def test_not_a_knot(self):
+        # derivatives at the boundaries
+        for i in range(k):
+            assert_allclose(b(self.xx[0], nu=i), b(self.xx[-1], nu=i), atol=1e-11)
+        # tests for axis=-1
```
How is this different from what's above?

ev-br commented on scipy/interpolate/tests/test_bsplines.py:
```diff
@@ -819,6 +820,93 @@ def test_not_a_knot(self):
+        x = np.sort(np.random.random_sample(n))
+        y = np.random.random_sample(n)
+        y[0] = y[-1] - 1  # to be sure that they are not equal
+        assert_raises(ValueError, make_interp_spline, x, y, k, None, 
```
Use 
```
with assert_raises(ValueError) :
    make_interp_spline(...) 
```

ev-br reviewed (changes requested):
It's getting closer. One more review round, algo review still pending.

ev-br reviewed (approved):
Looks good modulo small nits, which can taken care of in a sequel or not at all. 

Am planning to merge it soon (tm), unless there are more comments --- Would be nice to have a second pair of eyes on this BTW.

tylerjereddy commented:
> Looks good modulo small nits, which can taken care of in a sequel or not at all.

Alright, given that the only CI failure is unrelated I'm basically relying on this assessment and my own quick scan through the code for sanity checking and I'm going to merge it. It looks like @ev-br commented on an old test being deleted, but then resolved that conversation, so I'm assuming that is "ok"/approved here.

> Full disclosure: Egor and Diana work with me at HSE, so an additional review from someone else would be nice.

Agreed, but fairly specialized stuff, and we do have folks that appear to work at the same organization merging each others PRs without too much fuss already. Nice when we can get another person to check of course.

tylerjereddy commented:
Thanks @diana-sukhoverkhova @ev-br
